File: src/kotlin/app/adventofcode/impl/Day06.kt
package adventofcode.impl

import adventofcode.Helper
import adventofcode.Puzzle

class Day06(filename: String) : Puzzle(filename) {

    private val _daysToWatchFirstPuzzle = 80
    private val _daysToWatchSecondPuzzle = 256
    private val _cycle = 7
    private val _newFishOffset = 2
    private val _newFishValue = _cycle + _newFishOffset - 1
    private val _cycleRange = 1 until _cycle + _newFishOffset

    override fun solveFirst(): String? {
        var diagnostics = input.first().split(",").map { it.toInt() }
        Helper.debugMsg("Initial state: $diagnostics")

        val allFish = simulatePopulation(diagnostics, _daysToWatchFirstPuzzle)
        Helper.debugMsg("There are $allFish fish alive after $_daysToWatchFirstPuzzle days.")

        return "$allFish"
    }

    override fun solveSecond(): String? {
        var diagnostics = input.first().split(",").map { it.toInt() }
        Helper.debugMsg("Initial state: $diagnostics")

        val allFish = simulatePopulation(diagnostics, _daysToWatchSecondPuzzle)
        Helper.debugMsg("There are $allFish fish alive after $_daysToWatchSecondPuzzle days.")

        return "$allFish"
    }

    private fun simulatePopulation(initialPopulation: List<Int>, runTime: Int): Long {
        val fishCountPerCycle = LongArray(9)
        initialPopulation.forEach { fishCountPerCycle[it]++ }

        repeat(runTime) {
            val nNewFish = fishCountPerCycle[0]
            for (i in _cycleRange) {
                fishCountPerCycle[i - 1] = fishCountPerCycle[i]
            }
            fishCountPerCycle[_cycle - 1] += nNewFish
            fishCountPerCycle[_newFishValue] = nNewFish
        }

        return fishCountPerCycle.sum()
    }
}

File: src/csharp/impl/Day06.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Immutable;
using System.Text;
using static csharp.Helper;

namespace csharp.impl;

public class Day06 : IPuzzle
{
    private readonly string[] _input;

    private const int DaysToWatchFirstPuzzle = 80;
    private const int DaysToWatchSecondPuzzle = 256;
    private const short Cycle = 7;
    private const short NewFishOffset = 2;
    private const short ExtendedCycle = Cycle + NewFishOffset;

    public Day06(string filename)
    {
        _input = File.ReadAllLines(filename);
    }

    public string? FirstPuzzle()
    {
        var input = _input.SelectMany(line =>
                line.Split(",")
                    .Select(short.Parse))
            .ToList();

        DebugMsg($"Initial state: {string.Join(",", input)}");

        var fishCountPerCycle = new long[NewFishOffset];
        input.ForEach(fish => fishCountPerCycle[fish] += 1);

        for (var day = 0; day < DaysToWatchFirstPuzzle; day++)
        {
            var fishCountParents = fishCountPerCycle[0];
            for (var idx = 1; idx < ExtendedCycle; idx++)
            {
                fishCountPerCycle[idx - 1] = fishCountPerCycle[idx];
            }

            fishCountPerCycle[Cycle - 1] += fishCountParents;
            fishCountPerCycle[^1] = fishCountParents;
        }

        var result = fishCountPerCycle.Sum();
        DebugMsg($"There are {result} fish alive after {DaysToWatchFirstPuzzle} days.");
        return result.ToString();
    }

    public string? SecondPuzzle()
    {
        var input = _input.SelectMany(line =>
                line.Split(",")
                    .Select(short.Parse))
            .ToList();

        DebugMsg($"Initial state: {string.Join(",", input)}");

        var fishCountPerCycle = new long[NewFishOffset];
        input.ForEach(fish => fishCountPerCycle[fish] += 1);

        for (var day = 0; day < DaysToWatchSecondPuzzle; day++)
        {
            var fishCountParents = fishCountPerCycle[0];
            for (var idx = 1; idx < ExtendedCycle; idx++)
            {
                fishCountPerCycle[idx - 1] = fishCountPerCycle[idx];
            }

            fishCountPerCycle[Cycle - 1] += fishCountParents;
            fishCountPerCycle[^1] = fishCountParents;
        }

        var result = fishCountPerCycle.Sum();
        DebugMsg($"There are {result} fish alive after {DaysToWatchSecondPuzzle} days.");
        return result.ToString();
    }
}